### ISSUE TITLE
Keyboard shortcuts don't work on Wayland

### ISSUE THREAD

**Issue by Krzmbrzl** [mumble-voip/mumble#5257]
**Describe the bug**
When using Wayland instead of X11, the global shortcut engine does not work and as a consequence none of the user-registered shortcuts function.

**Steps to Reproduce**
1. Use Wayland
2. Try to use or register a shortcut

**Expected behavior**
Shortcuts to work normally

**Desktop (please complete the following information):**
 - OS: Linux (Wayland)
 - Version: all

**Additional context**
To my knowledge to this date Wayland (in contrast to X11) does not provide a way to either grab all user input nor does it allow to register global shortcuts. Instead one is supposed to register shortcuts with the currently used Window Manager (but there does not seem to exist a standard to do this).

Related issues: #3816 #3243 #4073

**Comment by davidebeatrici** [comment]
#3269

**Comment by icher** [comment]
Well, when I used 1.3.4 I can define hotkey for Push-to-talk and use it when Mumble window is focused. You can't use it in games whatever, but it's worked when you just want to talk. In 1.4 I can't do that. What should I do to use Push-to-talk when Mumble window is active at least?

**Comment by Krzmbrzl** [comment]
ATM there is no way around it. But there is already a ticket about weakening the disabling of shortcuts on Wayland.

**Comment by trthomps** [comment]
I was using PTT successfully in 1.3.4 as long as my mouse had focus on an X app (i.e. most games right now).  Would be nice to restore that behavior until global shortcuts are fixed in Wayland.

**Comment by zephrax** [comment]
Same here. It was working before, but with one of the last upgrades it seems that the functionality was disabled on purpose. I was able to use it successfully while focusing on a X Window.

**Comment by Nothankslads** [comment]
Has there been any progress on this?

**Comment by Krzmbrzl** [comment]
No

**Comment by jsparber** [comment]
> What should I do to use Push-to-talk when Mumble window is active at least?

@icher You can enable "Display push to talk window" to get a extra window with a button. It's not great but better then nothing.

![image](https://user-images.githubusercontent.com/1911029/168845364-ed673c19-cc9a-49a4-bb3a-25b4d74a54b2.png)

**Comment by icher** [comment]
> @icher You can enable "Display push to talk window" to get a extra window with a button. It's not great but better then nothing.

Nah, it's not usable. I just downgrade to 1.3.4 and run as `QT_QPA_PLATFORM=xcb mumble`.

**Comment by bovinespongiformflu** [comment]
Please restore old functionality, having it disabled entirely isnt helping anyone and it still worked fine in proton/x games.

**Comment by tidux** [comment]
A workaround is to use the RPC client to map keys to commands in your WM/DE, although you do need to dedicate two keybinds for it:

`mumble rpc starttalking` and `mumble rpc stoptalking` do what you'd expect for PTT.

**Comment by travismorton** [comment]
I've been running a hacky workaround using the mumble D-Bus API and `libinput`. As long as your user is part of the `input` group you can listen for mouse/keyboard down/up events and forward them to the `startTalking` and `stopTalking` D-Bus methods. Here's the [script](https://gist.github.com/travismorton/aef78d9f25c690baa769b9110ee896a6) I run in the background. It will enable PTT when BTN_EXTRA is pushed. 

Before this I was using Sway's `bindsym` to emit the D-Bus calls similar to what @tidux suggested, but events would get dropped based on where the mouse was. Going directly to `libinput` helped me avoid the dropped button-up events which caused state issues and an always-on mic sometimes. It also doesn't depend on focus being on an X window for PTT to work.

**Comment by carlocastoldi** [comment]
i love your script, @travismorton. Thanks a lot!

I think it removes the security of Wayland from possible keyloggers, but it's a price i'm ready to pay. Even more as I expect it to be a temporary solution. At least until a shortcuts xdg-portal is not finalized

**Comment by travismorton** [comment]
Good point, I should have mentioned the possible security hole in adding your user to the `input` group. There's more nuanced ways to do that which involve udev rules and logind as mentioned in [this](https://old.reddit.com/r/linuxquestions/comments/bh4ex1/is_adding_a_user_to_input_group_secure/elq3uvk/) reddit thread, although they don't completely solve the issue, just limit the blast radius.

After you mentioned pinning the event path I changed the script to fetch the context based on `libinput` seat rather than path. Assuming your DE/WM doesn't create multiple seats this should always be `seat0` (as far as my limited understanding goes) and shouldn't change. The gist is updated as well, hopefully that accomplishes what you're looking for?

**Comment by carlocastoldi** [comment]
> here's more nuanced ways to do that which involve udev rules and logind as mentioned in [this](https://old.reddit.com/r/linuxquestions/comments/bh4ex1/is_adding_a_user_to_input_group_secure/elq3uvk/) reddit thread

I was indeed wondering if we couldn't do some udev rule magic!

> although they don't completely solve the issue, just limit the blast radius

it limits the "keylogging capabilities" to the device you want to bind PTT to, right? Or is the blast radius wider?

> The gist is updated as well, hopefully that accomplishes what you're looking for?

It does! It's awesome, thanks!

**Comment by ananace** [comment]
There is the XDG global shortcut portal API ([PR](https://github.com/flatpak/xdg-desktop-portal/pull/711), [Docs](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.GlobalShortcuts)) which should also allow for binding global shortcuts on Wayland in a DE-agnostic way - though with a bit of DBus work, unfortunately it's not actually available in any released version of KDE/GNOME yet, so it'll probably be a while until it can be used.

**Comment by RichardFevrier** [comment]
Will it be possible to use only one key to bind the Push To Talk functionality?

**Comment by RichardFevrier** [comment]
I meant; will the xdg-desktop-portal GlobalShortcuts let's mumble users to use only a single key for the Push to Talk functionality when everything will be implemented?

**Comment by Krzmbrzl** [comment]
Ah :bulb: 
Yes I would expect so. The hope would be for this change to be completely transparent such that the user doesn't notice whether they are using Mumble with X11 or with Wayland.

**Comment by RichardFevrier** [comment]
Maybe @aleixpol would know

**Comment by aleixpol** [comment]
Yes, it's possible to use a single key for push to talk as does this PR, it's not something theoretical.

**Comment by jp-bennett** [comment]
KDE has added "Legacy X11 App Support" to the system settings when running under Wayland, and one of the options is to allow applications to read keystrokes again. This completely fixes push-to-talk support on Fedora. There is a merge request (https://invent.kde.org/plasma/kwin/-/merge_requests/4595) to extend this system to also send mouse button presses.

**Comment by CmdrMoozy** [comment]
Note that for now at least the "Legacy X11 App Support" only seems to work for keyboard shortcuts. For mouse button support, it appears KDE 6.1 will be needed: https://www.mail-archive.com/kde-bugs-dist@kde.org/msg889735.html

**Comment by Mikilio** [comment]
I use hyprland and in a very wayland like way I want to let the compositor handle global shortcuts for me by sending it to the mumble application. I believe that also in the future, global shortcuts should be handled by the compositor. Maybe we can ask freedesktop to make a standard configuration format for this. That way, mumble can suggest configurations that work independent of the compositor. This way there are no security compromises as the user will be conscious of all changes.

**Comment by Ases** [comment]
> Note that for now at least the "Legacy X11 App Support" only seems to work for keyboard shortcuts. For mouse button support, it appears KDE 6.1 will be needed: https://www.mail-archive.com/kde-bugs-dist@kde.org/msg889735.html

I confirm that the Push To Talk with a mouse button is now functional under KDE Plasma 6.1 configuring the Legacy X11 App Support. :tada:

**Comment by BadCo-NZ** [comment]
> > Note that for now at least the "Legacy X11 App Support" only seems to work for keyboard shortcuts. For mouse button support, it appears KDE 6.1 will be needed: https://www.mail-archive.com/kde-bugs-dist@kde.org/msg889735.html
> 
> I confirm that the Push To Talk with a mouse button is now functional under KDE Plasma 6.1 configuring the Legacy X11 App Support. 🎉

How did you do this? Mumble flatpak crashes for me on Fedora Kinoite when I enable Global Shortcuts.

**Comment by Ases** [comment]
> > > Note that for now at least the "Legacy X11 App Support" only seems to work for keyboard shortcuts. For mouse button support, it appears KDE 6.1 will be needed: https://www.mail-archive.com/kde-bugs-dist@kde.org/msg889735.html
> > 
> > 
> > I confirm that the Push To Talk with a mouse button is now functional under KDE Plasma 6.1 configuring the Legacy X11 App Support. 🎉
> 
> How did you do this? Mumble flatpak crashes for me on Fedora Kinoite when I enable Global Shortcuts.


I'm currently using Mumble v1.5.517 from snapcrafters on KDE neon. 
Although I'm not a fan of snaps and generally avoid them, I wanted
to try it.
I then simply configured "Legacy X11 App Support" without any special adjustments.

**Comment by trthomps** [comment]
I can also confirm it's been working for a while on Arch Linux with the system packages.

**Comment by aleixpol** [comment]
Please don't consider the issue as addressed with the "Legacy X11 app support". Because it's only for X11 and effectively a backwards-compatibility workaround, not the right way to support it on Wayland.
https://github.com/mumble-voip/mumble/pull/5976

**Comment by dotbanana** [comment]
So after all those years this is still not working yes?  
Is mumble dead?

**Comment by omnigenous** [comment]
@Mikilio I'm also on hyprland, did you find a workaround?

**Comment by IDDQD69** [comment]
> @Mikilio I'm also on hyprland, did you find a workaround?

Hyprland works just fine for me:

input {
      ...
      kb_options = caps:none

}

bind = , VoidSymbol, pass, ^(info\.mumble\.Mumble)$
bind = , VoidSymbol, pass, ^(discord)$

worked just fine

**Comment by omnigenous** [comment]
@IDDQD69 this is really bizarre, I have same key (`F10`) to mute mic both in mumble and teamspeak. So when teamspeak is open along side mumble `F10` to mute mic works in both teamspeak and mumble and I even removed global pass with mumbles class and it still worked with this for mumble:

```
bind = , F10, pass, ^(TeamSpeak)$
```

So apparently global keybind is leaking to other apps beside teamspeak unless I'm missing something.

Do you know if teamspeak starting something in background to make global keybind work? I want to be able to have only mumble open and use mute keybind.

**Comment by powerjungle** [comment]
> @icher You can enable "Display push to talk window" to get a extra window with a button. It's not great but better then nothing.

Or you can run Mumble like this to force it into xwayland: `env XDG_SESSION_TYPE=x11 mumble`

**Comment by danielwerg** [comment]
I'm on hyprland and also has this issue. I only needed to mute and deafen so thankfully to mumble rpc I could bind those commands instead of using [global] keybinds!

```sh
mumble rpc togglemute
mumble rpc toggledeaf
```

**Comment by Fifthdread** [comment]
I run the Flatpack version of Mumble and I had to create the custom kde command to get a global shortcut to work for me, as well as force Mumble to launch via x11.

`flatpak run info.mumble.Mumble rpc togglemute`

... jank, but it works.

**Comment by julien9808** [comment]
with Gnome, disable Wayland windowing system works in Flatseal works for  me too

**Comment by steveyh25** [comment]
making the app run as x11 and enabling legacy X11 support makes mumble work again, but I've noticed that discord does not have this problem and runs as a wayland app so a solution must be possible

**Comment by Krzmbrzl** [comment]
> a solution must be possible

It's possible. We "only" need to use the xdg-portal way of doing shortcuts, which is the intended way of handling global shortcuts under Wayland. The problem is that this is a very much non-trivial change for which we haven't found the time yet. There is #5976 but the same things apply to that: not enough resources have been found for getting it done.

**Comment by kooltux** [comment]
Maybe already mentioned somewhere, but this is something I tested today successfully on openSUSE Leap 16.0 : defining a shortcut in Gnome keyboard settings (settings->keyboard->view and customize shortcuts->custom shortcuts)  to run `/usr/bin/mumble rpc togglemute`

**Comment by steveyh25** [comment]
it appears as though later versions of fedora require you to force x11 via flatpak configuration (and not something like flatseal), otherwise the application crashes on startup:

    flatpak override --user --socket=x11 info.mumble.Mumble

**Comment by WereCatf** [comment]
In KDE Plasma6, you can go to Settings -> Application Permissions and allow X11-apps to listen to keyboard- and mouse-events like they used to work and now push-to-talk works like one would expect. No need for any rpc toggles or scripts.

Obviously, proper Wayland-support would be nice, but at least this way Mumble is perfectly useable.

**Comment by steveyh25** [comment]
> In KDE Plasma6, you can go to Settings -> Application Permissions and allow X11-apps to listen to keyboard- and mouse-events like they used to work and now push-to-talk works like one would expect. No need for any rpc toggles or scripts.
> 
> Obviously, proper Wayland-support would be nice, but at least this way Mumble is perfectly useable.

yes... legacy x11 support...

unfortunately, if you use a mouse button as your push-to-talk then you will run into the next issue, which is that KDE currently has a bug where pressing the mouse button will randomly change window focus

mumble is the only app that I need to enable legacy x11 support for

**Comment by devinsaw** [comment]
> In KDE Plasma6, you can go to Settings -> Application Permissions and allow X11-apps to listen to keyboard- and mouse-events like they used to work and now push-to-talk works like one would expect. No need for any rpc toggles or scripts.
> 
> Obviously, proper Wayland-support would be nice, but at least this way Mumble is perfectly useable.

I've tried this but the mumble client still crashes when trying to add a PTT shortcut inside mumble.

**Comment by plasmatoasthub** [comment]
if you are using hyprland which is Wayland compositor then it works perfectly fine unless you are using flatpak version... u need to install it with sudo pacman -S mumble
flatpack run on pure wayland

**Comment by danielwerg** [comment]
@plasmatoasthub could u provide more details about ur setup? i can't even assign the key.

<img width="492" height="193" alt="Image" src="https://github.com/user-attachments/assets/67b44075-aa4a-4543-818f-19b197856bb9" />

**Comment by jadrienb** [comment]
solved with this 

<img width="1663" height="979" alt="Image" src="https://github.com/user-attachments/assets/20b2ff21-1a49-4035-90ce-eea7ea156076" />

**Comment by distinctjuggle** [comment]
How can one force legacy X11 mode with Mumble under a Wayland session (not using flatpak)?  I've tried a few environment variables, but they don't seem to work per others results here under Plasma/KDE:

```
mumble --socket=x11
GDK_BACKEND=x11 mumble
QT_QPA_PLATFORM=xcb mumble
mumble -platform xcb
```

I'm new to Wayland, so forgive my ignorance if it's a common thing.  I can't seem to find a working answer here.

**Comment by WereCatf** [comment]
> How can one force legacy X11 mode with Mumble under a Wayland session (not using flatpak)? I've tried a few environment variables, but they don't seem to work per others results here under Plasma/KDE

It has been mentioned several times in the comments above yours: just go to KDE settings, then to Application permissions.

**Comment by distinctjuggle** [comment]
> > How can one force legacy X11 mode with Mumble under a Wayland session (not using flatpak)? I've tried a few environment variables, but they don't seem to work per others results here under Plasma/KDE
> 
> It has been mentioned several times in the comments above yours: just go to KDE settings, then to Application permissions.

I didn't ask for how to enable the KDE/Plasma setting to enable global shortcuts for KDE/Plasma; I asked how to launch mumble in X11 compatibility mode instead of using Wayland.  I even provided examples of how I'm trying to do this, clearly demonstrating the difference in goal.

**Comment by WereCatf** [comment]
> > > How can one force legacy X11 mode with Mumble under a Wayland session (not using flatpak)? I've tried a few environment variables, but they don't seem to work per others results here under Plasma/KDE
> > 
> > 
> > It has been mentioned several times in the comments above yours: just go to KDE settings, then to Application permissions.
> 
> I didn't ask for how to enable the KDE/Plasma setting to enable global shortcuts for KDE/Plasma; I asked how to launch mumble in X11 compatibility mode instead of using Wayland. I even provided examples of how I'm trying to do this, clearly demonstrating the difference in goal.

Well, this entire thread is about the keyboard shortcuts. If that's not what you're here for then you're off-topic.

**Comment by distinctjuggle** [comment]
> > > > How can one force legacy X11 mode with Mumble under a Wayland session (not using flatpak)? I've tried a few environment variables, but they don't seem to work per others results here under Plasma/KDE
> > > 
> > > 
> > > It has been mentioned several times in the comments above yours: just go to KDE settings, then to Application permissions.
> > 
> > 
> > I didn't ask for how to enable the KDE/Plasma setting to enable global shortcuts for KDE/Plasma; I asked how to launch mumble in X11 compatibility mode instead of using Wayland. I even provided examples of how I'm trying to do this, clearly demonstrating the difference in goal.
> 
> Well, this entire thread is about the keyboard shortcuts. If that's not what you're here for then you're off-topic.

If you don't understand what I'm asking, that's okay.  You don't have to respond to a question you don't understand.

I'll explain for you anyway.

This workaround which you pointed me to works for applications running using some X11 compatibility layer under a Wayland session.  If Mumble is launched using Wayland without this compatibility tool, then the KDE/Plasma setting for Legacy X11 application global hotkey support isn't going to apply to Mumble.  The above users are enabling X11 on Wayland using flatpak.  I was inquiring on how to do this without flatpak so that the KDE/Plasma setting becomes applicable.  As I already explained in my original comment, the KDE/Plasma setting isn't working for me, and I'm presuming this is because Mumble is just launching under normal Wayland (I could be wrong on this point, though.  I don't know a 100% foolproof way to test, other than that I'm seeing the Wayland specific warning under the Mumble shortcut/keybind setting page, which I would presume is only present under Wayland).

This question is **explicitly** about the problem discussed in this issue.

**Comment by WereCatf** [comment]
> > > > > How can one force legacy X11 mode with Mumble under a Wayland session (not using flatpak)? I've tried a few environment variables, but they don't seem to work per others results here under Plasma/KDE
> > > > 
> > > > 
> > > > It has been mentioned several times in the comments above yours: just go to KDE settings, then to Application permissions.
> > > 
> > > 
> > > I didn't ask for how to enable the KDE/Plasma setting to enable global shortcuts for KDE/Plasma; I asked how to launch mumble in X11 compatibility mode instead of using Wayland. I even provided examples of how I'm trying to do this, clearly demonstrating the difference in goal.
> > 
> > 
> > Well, this entire thread is about the keyboard shortcuts. If that's not what you're here for then you're off-topic.
> 
> If you don't understand what I'm asking, that's okay. You don't have to respond to a question you don't understand.
> 
> I'll explain for you anyway.
> 
> This workaround which you pointed me to works for applications running using some X11 compatibility layer under a Wayland session. If Mumble is launched using Wayland without this compatibility tool, then the KDE/Plasma setting for Legacy X11 application global hotkey support isn't going to apply to Mumble. The above users are enabling X11 on Wayland using flatpak. I was inquiring on how to do this without flatpak so that the KDE/Plasma setting becomes applicable.
> 
> This question is **explicitly** about the problem discussed in this issue.

Wow, someone gets upset pretty easily. I'll let you know that I am not using Mumble flatpak and the KDE-settings work just fine without extra work or "compatibility tool"

**Comment by distinctjuggle** [comment]
> > > > > > How can one force legacy X11 mode with Mumble under a Wayland session (not using flatpak)? I've tried a few environment variables, but they don't seem to work per others results here under Plasma/KDE
> > > > > 
> > > > > 
> > > > > It has been mentioned several times in the comments above yours: just go to KDE settings, then to Application permissions.
> > > > 
> > > > 
> > > > I didn't ask for how to enable the KDE/Plasma setting to enable global shortcuts for KDE/Plasma; I asked how to launch mumble in X11 compatibility mode instead of using Wayland. I even provided examples of how I'm trying to do this, clearly demonstrating the difference in goal.
> > > 
> > > 
> > > Well, this entire thread is about the keyboard shortcuts. If that's not what you're here for then you're off-topic.
> > 
> > 
> > If you don't understand what I'm asking, that's okay. You don't have to respond to a question you don't understand.
> > I'll explain for you anyway.
> > This workaround which you pointed me to works for applications running using some X11 compatibility layer under a Wayland session. If Mumble is launched using Wayland without this compatibility tool, then the KDE/Plasma setting for Legacy X11 application global hotkey support isn't going to apply to Mumble. The above users are enabling X11 on Wayland using flatpak. I was inquiring on how to do this without flatpak so that the KDE/Plasma setting becomes applicable.
> > This question is **explicitly** about the problem discussed in this issue.
> 
> Wow, someone gets upset pretty easily. I'll let you know that I am not using Mumble flatpak and the KDE-settings work just fine without extra work or "compatibility tool"

I'm really not sure how anyone could think I was upset based on my comments here.  I just responded to your statements...?

My point to you was pretty simple - before you tell others that they're off-topic, you should probably understand what they're asking.  I may very well be misunderstanding the situation with Wayland/X11 here, but I was clearly asking about launching Mumble in an X11 compatibility mode (eg xwayland) for the purpose of functional keyboard shortcuts.  Seemingly this is necessary, presuming Mumble natively handles Wayland (I believe it does from the settings dialog/hint) and the KDE/Plasma Legacy X11 app page only works for Legacy X11 apps.  Again, I could be wrong on this point (maybe Mumble only has an X11 version still?  But the dialog only appearing on Wayland, and shortcuts disabling themselves when launching on Wayland would indicate otherwise to me).

Regardless of that, telling me to just, "read above" and "you're off-topic" is not helpful when you aren't reading (or at the very least, trying to understand) what I've wrote.  By all means we can engage in a conversation to clarify if I was somehow ambiguous (I really don't think I was), but that isn't really where you went with that.

Best wishes and no hard feelings from me.  I'm just going to use the RPC workaround until proper Wayland support gets added I guess.

**Comment by WereCatf** [comment]
> Seemingly this is necessary, presuming Mumble natively handles Wayland (I believe it does from the settings dialog/hint) and the KDE/Plasma Legacy X11 app page only works for Legacy X11 apps. Again, I could be wrong on this point.

The X11-socket thing is just for the Flatpak version, not for the Mumble installed from one's distro's repos. The KDE-settings do apply.
